tensorboard/plugins/text_v2/text_demo_v2.py

```diff
@@ -0,0 +1,146 @@
+    # tables. Markdown tables look like this:
+    #
+    #     | hello | there |
+    #     |-------|-------|
+    #     | this  | is    |
+    #     | a     | table |
+    #
+    # The leading and trailing pipes in each row are optional, and the text
+    # doesn't actually have to be neatly aligned, so we can create these
+    # pretty easily. Let's do so.
+    header_row = "Pounds of chocolate | Happiness"
+    chocolate = tf.range(step)
+    happiness = tf.square(chocolate + 1)
+    chocolate_column = tf.as_string(chocolate)
+    happiness_column = tf.as_string(happiness)
+    table_rows = tf.strings.join([chocolate_column, " | ", happiness_column])
+    table_body = tf.strings.reduce_join(inputs=table_rows, separator="\n")
+    table = tf.strings.join([header_row, "---|---", table_body], separator="\n")
+    preamble = "We conducted an experiment and found the following data:\n\n"
+    result = tf.strings.join([preamble, table])
+    text("chocolate_study", result, step)
+
+
+def higher_order_tensors(step):
+    # We're not limited to passing scalar tensors to the summary
+    # operation. If we pass a rank-1 or rank-2 tensor, it'll be visualized
+    # as a table in TensorBoard. (For higher-ranked tensors, you'll see
+    # just a 2D slice of the data.)
+    #
+    # To demonstrate this, let's create a multiplication table.
+
+    # First, we'll create the table body, a `step`-by-`step` array of
+    # strings.
+    numbers = tf.range(step)
+    numbers_row = tf.expand_dims(numbers, 0)  # shape: [1, step]
+    numbers_column = tf.expand_dims(numbers, 1)  # shape: [step, 1]
+    products = tf.matmul(numbers_column, numbers_row)  # shape: [step, step]
+    table_body = tf.as_string(products)
+
+    # Next, we'll create a header row and column, and a little
+    # multiplication sign to put in the corner.
+    bold_numbers = tf.strings.join(["**", tf.as_string(numbers), "**"])
+    bold_row = tf.expand_dims(bold_numbers, 0)
+    bold_column = tf.expand_dims(bold_numbers, 1)
+    corner_cell = tf.constant(u"\u00d7".encode("utf-8"))  # MULTIPLICATION SIGN
+
+    # Now, we have to put the pieces together. Using `axis=0` stacks
+    # vertically; using `axis=1` juxtaposes horizontally.
+    table_body_and_top_row = tf.concat([bold_row, table_body], axis=0)
+    table_left_column = tf.concat([[[corner_cell]], bold_column], axis=0)
+    table_full = tf.concat([table_left_column, table_body_and_top_row], axis=1)
+
+    # The result, `table_full`, is a rank-2 string tensor of shape
+    # `[step + 1, step + 1]`. We can pass it directly to the summary, and
+    # we'll get a nicely formatted table in TensorBoard.
+    text("multiplication_table", table_full, step)
+
+
+def create_run(logdir, run_name):
+    # Creates a run specified by `run_name` populated with the text summaries
+    # generated by the function above.
+    writer = tf.summary.create_file_writer(logdir + "/" + run_name)
```

Comment:
Question to @nfelt: it seems like demos are hard coding logdir like `/tmp/images_demo` which probably won't work on Windows. Is there a reason why we did that instead of using https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp and `os.path.join`?

Reply:
I missed this question from before.  It would be better if these worked on windows, but using `mkdtemp` would have the disadvantage that the directory is no longer predictable on posix (makes it a bit harder to do things like run the demo and then open TensorBoard pointed at the logdir).  So might be nice to keep the `/tmp/foo_demo` naming structure for that reason, but we could use `mkdtemp` on Windows?

Reply:
Original author here. The purpose was indeed to put the output in a
predictable location. There was also a plan to have a single “mega-demo”
target that would have run all the demos and put the output in a common
directory, hence the common `run(logdir)` structure to the demos, but we
never got around to that. I’m not inclined to go out of our way to
support Windows for _development_ processes, rather than end users,
given that none of us actually uses it and we’ve had minimal demand.